vllm/v1/core/scheduler.py

```diff
@@ -480,23 +518,47 @@ def update_from_output(
             scheduler_stats=self.make_stats(),
         )
 
-    def _check_stop(self, request: Request) -> bool:
+    def _maybe_stop_and_crop(self, request: Request) -> bool:
+        """Check if the request should be stopped.
+        The function should handle both single token generation or
+        multiple token generation (e.g., spec decode) per step.
+        
+        This function will crop requests because the request is stopped
+        in the middle of the generation.
+        When cropping, we do not need to update the input batch because
+        it will be updated in the next execute_model call's
+        _update_states method, where the request data is aligned
+        with the data in the persistent batch.
+        """
         if (request.num_tokens >= self.max_model_len
                 or request.num_output_tokens >= request.max_tokens):
             request.status = RequestStatus.FINISHED_LENGTH_CAPPED
+            num_total_token = min(
+                self.max_model_len, request.num_tokens,
+                request.max_tokens + request.num_prompt_tokens,
+                request.num_output_tokens + request.num_prompt_tokens)
+            request.crop(num_total_token)
             return True
 
         sampling_params = request.sampling_params
-        last_token_id = request.output_token_ids[-1]
-        if (not sampling_params.ignore_eos
-                and last_token_id == request.eos_token_id):
-            request.status = RequestStatus.FINISHED_STOPPED
-            return True
-
-        if last_token_id in (sampling_params.stop_token_ids or ()):
-            request.status = RequestStatus.FINISHED_STOPPED
-            request.stop_reason = last_token_id
-            return True
+        if not sampling_params.ignore_eos:
+            assert request.eos_token_id is not None
+            if request.eos_token_id in request.output_token_ids:
+                request.status = RequestStatus.FINISHED_STOPPED
+                num_total_token = request.num_prompt_tokens + \
+                        request.output_token_ids.index(request.eos_token_id) + 1
+                request.crop(num_total_token)
+                return True
+
+        output_token_ids = set(request.output_token_ids)
+        for stop_token_id in sampling_params.stop_token_ids:
+            if stop_token_id in output_token_ids:
+                request.status = RequestStatus.FINISHED_STOPPED
+                request.stop_reason = stop_token_id
+                num_total_token = request.num_prompt_tokens + \
+                        request.output_token_ids.index(stop_token_id) + 1
+                request.crop(num_total_token)
+                return True
```

Comment:
Add a comment and TODO about removing this loop when we support stop checking in the proposer so that the stop token must be the last generated token.